vm/interpreter/src/fvm/externs.rs

```diff
@@ -231,3 +222,89 @@ impl<DB: BlockStore> Consensus for ForestExterns<DB> {
         Ok((cf, total_gas))
     }
 }
+
+fn cal_gas_used_from_stats(
+    stats: Ref<BSStats>,
+    network_version: NetworkVersion,
+) -> anyhow::Result<Gas> {
+    let price_list = price_list_by_network_version(network_version);
+    let mut gas_tracker = GasTracker::new(Gas::new(i64::MAX), Gas::new(0));
+    for _ in 0..stats.r {
+        gas_tracker.apply_charge(price_list.on_block_open_base())?;
+    }
+    if stats.w > 0 {
+        gas_tracker.apply_charge(price_list.on_block_link(stats.bw))?;
+        for _ in 1..stats.w {
+            gas_tracker.apply_charge(price_list.on_block_link(0))?;
+        }
+    }
+    Ok(gas_tracker.gas_used())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use anyhow::ensure;
+    use std::{cell::RefCell, iter::repeat};
+
+    #[test]
+    fn test_cal_gas_used_from_stats_1_read() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(1, &[])
+    }
+
+    #[test]
+    fn test_cal_gas_used_from_stats_1_write() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(0, &[100])
+    }
+
+    #[test]
+    fn test_cal_gas_used_from_stats_multi_read() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(10, &[])
+    }
+
+    #[test]
+    fn test_cal_gas_used_from_stats_multi_write() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(0, &[100, 101, 102, 103, 104, 105, 106, 107, 108, 109])
+    }
+
+    #[test]
+    fn test_cal_gas_used_from_stats_1_read_1_write() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(1, &[100])
+    }
+
+    #[test]
+    fn test_cal_gas_used_from_stats_multi_read_multi_write() -> anyhow::Result<()> {
+        test_cal_gas_used_from_stats_inner(10, &[100, 101, 102, 103, 104, 105, 106, 107, 108, 109])
+    }
+
+    fn test_cal_gas_used_from_stats_inner(
+        read_count: usize,
+        write_bytes: &[usize],
+    ) -> anyhow::Result<()> {
+        let network_version = NetworkVersion::V8;
+        let stats = BSStats {
+            r: read_count,
+            w: write_bytes.len(),
+            br: 0, // Not used in current logic
+            bw: write_bytes.iter().sum(),
+        };
+        println!("stats: {:?}", stats);
```

Comment:
do we need this print? I understand it's for debugging purposes only?

Reply:
fixed